File: src/c-sharp/NotebookAutomation.Core/Tools/TagManagement/TagService.cs
// Licensed under the MIT License. See LICENSE file in the project root for full license information.

using NotebookAutomation.Core.Utils;

namespace NotebookAutomation.Core.Tools.TagManagement;

/// <summary>
/// Implementation of <see cref="ITagService"/> that wraps <see cref="TagProcessor"/>.
/// </summary>
public class TagService : ITagService
{
    private readonly ILogger<TagService> _logger;
    private readonly ILoggerFactory _loggerFactory;
    private readonly IYamlHelper _yamlHelper;
    private readonly IMetadataSchemaLoader? _schemaLoader;
    private readonly string _vaultRootPath;

    /// <summary>
    /// Initializes a new instance of the <see cref="TagService"/> class.
    /// </summary>
    /// <param name="logger">Logger for service operations.</param>
    /// <param name="loggerFactory">Factory for creating TagProcessor loggers.</param>
    /// <param name="yamlHelper">YAML helper for frontmatter processing.</param>
    /// <param name="schemaLoader">Optional metadata schema loader.</param>
    /// <param name="vaultRootPath">Root path of the Obsidian vault.</param>
    public TagService(
        ILogger<TagService> logger,
        ILoggerFactory loggerFactory,
        IYamlHelper yamlHelper,
        IMetadataSchemaLoader? schemaLoader,
        string vaultRootPath)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
        _yamlHelper = yamlHelper ?? throw new ArgumentNullException(nameof(yamlHelper));
        _schemaLoader = schemaLoader;
        _vaultRootPath = vaultRootPath ?? throw new ArgumentNullException(nameof(vaultRootPath));
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> AddNestedTagsAsync(string path, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Adding nested tags to {Path}, DryRun={DryRun}", path, dryRun);

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);

            Dictionary<string, int> stats;
            if (Directory.Exists(resolvedPath))
            {
                stats = await processor.ProcessDirectoryAsync(resolvedPath);
            }
            else if (File.Exists(resolvedPath))
            {
                var success = await processor.ProcessFileAsync(resolvedPath);
                stats = processor.Stats;
            }
            else
            {
                return CreateErrorResult($"Path not found: {resolvedPath}", dryRun);
            }

            return CreateResult(stats, dryRun, "Added nested tags");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error adding nested tags to {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> ConsolidateTagsAsync(string? path = null, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Consolidating tags in {Path}, DryRun={DryRun}", path ?? "vault", dryRun);

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);

            // Consolidate uses the same processing as add-nested but focuses on tag deduplication
            var stats = await processor.ProcessDirectoryAsync(resolvedPath);

            return CreateResult(stats, dryRun, "Consolidated tags");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error consolidating tags in {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> RestructureTagsAsync(string? path = null, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Restructuring tags in {Path}, DryRun={DryRun}", path ?? "vault", dryRun);

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);
            var stats = await processor.RestructureTagsInDirectoryAsync(resolvedPath);

            return CreateResult(stats, dryRun, "Restructured tags");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error restructuring tags in {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> UpdateFrontmatterAsync(string path, string key, string value, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Updating frontmatter {Key}={Value} in {Path}, DryRun={DryRun}", key, value, path, dryRun);

        if (string.IsNullOrWhiteSpace(key))
        {
            return CreateErrorResult("Key cannot be empty", dryRun);
        }

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);
            var stats = await processor.UpdateFrontmatterKeyAsync(resolvedPath, key, value);

            return CreateResult(stats, dryRun, $"Updated frontmatter key '{key}'");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating frontmatter in {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    /// <inheritdoc/>
    public async Task<YamlDiagnosisResult> DiagnoseYamlAsync(string? path = null, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Diagnosing YAML in {Path}", path ?? "vault");

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return new YamlDiagnosisResult
                {
                    Success = false,
                    Message = $"Invalid path: {validationError}",
                    FilesScanned = 0,
                    FilesWithIssues = 0
                };
            }

            var processor = CreateTagProcessor(dryRun: true, verbose: false);
            var results = await processor.DiagnoseFrontmatterIssuesAsync(resolvedPath);

            var issues = results.Select(r => new YamlIssue
            {
                FilePath = r.FilePath,
                Description = r.DiagnosticMessage,
                SuggestedFix = GetSuggestedFix(r.DiagnosticMessage)
            }).ToList();

            // Count files scanned by counting markdown files
            var filesScanned = Directory.Exists(resolvedPath)
                ? Directory.GetFiles(resolvedPath, "*.md", SearchOption.AllDirectories).Length
                : 1;

            return new YamlDiagnosisResult
            {
                Success = true,
                Message = issues.Count == 0
                    ? "No YAML issues found"
                    : $"Found {issues.Count} file(s) with YAML issues",
                FilesScanned = filesScanned,
                FilesWithIssues = issues.Count,
                Issues = issues
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error diagnosing YAML in {Path}", path);
            return new YamlDiagnosisResult
            {
                Success = false,
                Message = $"Error: {ex.Message}",
                FilesScanned = 0,
                FilesWithIssues = 0
            };
        }
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> CheckMetadataAsync(string? path = null, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Checking metadata in {Path}, DryRun={DryRun}", path ?? "vault", dryRun);

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);
            var stats = await processor.CheckAndEnforceMetadataConsistencyAsync(resolvedPath);

            return CreateResult(stats, dryRun, "Checked metadata consistency");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking metadata in {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    /// <inheritdoc/>
    public async Task<TagOperationResult> CleanIndexFilesAsync(string? path = null, bool dryRun = false, CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("Cleaning index files in {Path}, DryRun={DryRun}", path ?? "vault", dryRun);

        try
        {
            var resolvedPath = ResolvePath(path);
            if (!ValidatePath(resolvedPath, out var validationError))
            {
                return CreateErrorResult($"Invalid path: {validationError}", dryRun);
            }

            var processor = CreateTagProcessor(dryRun, verbose: false);

            // Find all index files
            var indexFiles = new List<string>();
            if (Directory.Exists(resolvedPath))
            {
                indexFiles.AddRange(Directory.GetFiles(resolvedPath, "_index.md", SearchOption.AllDirectories));
                indexFiles.AddRange(Directory.GetFiles(resolvedPath, "index.md", SearchOption.AllDirectories));
            }
            else if (File.Exists(resolvedPath) && IsIndexFile(resolvedPath))
            {
                indexFiles.Add(resolvedPath);
            }

            var filesProcessed = 0;
            var filesModified = 0;
            var filesWithErrors = 0;

            foreach (var indexFile in indexFiles)
            {
                filesProcessed++;
                try
                {
                    var content = await File.ReadAllTextAsync(indexFile, cancellationToken);
                    var frontmatterYaml = _yamlHelper.ExtractFrontmatter(content);
                    if (string.IsNullOrWhiteSpace(frontmatterYaml))
                    {
                        continue; // No frontmatter to process
                    }

                    var frontmatter = _yamlHelper.ParseYamlToDictionary(frontmatterYaml);
                    var body = _yamlHelper.RemoveFrontmatter(content);

                    var success = await processor.ClearTagsFromFileAsync(indexFile, frontmatter, body);
                    if (success)
                    {
                        filesModified++;
                    }
                }
                catch (Exception ex)
                {
                    _logger.LogWarning(ex, "Error processing index file {File}", indexFile);
                    filesWithErrors++;
                }
            }

            return new TagOperationResult
            {
                Success = filesWithErrors == 0,
                Message = $"Cleaned {filesModified} of {filesProcessed} index files",
                FilesProcessed = filesProcessed,
                FilesModified = filesModified,
                TagsAdded = 0,
                FilesWithErrors = filesWithErrors,
                DryRun = dryRun
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error cleaning index files in {Path}", path);
            return CreateErrorResult($"Error: {ex.Message}", dryRun);
        }
    }

    #region Private Helpers

    private TagProcessor CreateTagProcessor(bool dryRun, bool verbose)
    {
        var tagProcessorLogger = _loggerFactory.CreateLogger<TagProcessor>();
        var failedLogger = _loggerFactory.CreateLogger("FailedOperations");

        return new TagProcessor(
            tagProcessorLogger,
            failedLogger,
            _yamlHelper,
            dryRun,
            verbose,
            resolverRegistry: _schemaLoader?.ResolverRegistry);
    }

    private string ResolvePath(string? path)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            return _vaultRootPath;
        }

        // If it's already an absolute path, return it
        if (Path.IsPathRooted(path))
        {
            return PathUtils.NormalizePath(path);
        }

        // Resolve relative path against vault root
        return PathUtils.NormalizePath(Path.Combine(_vaultRootPath, path));
    }

    private bool ValidatePath(string resolvedPath, out string? error)
    {
        error = null;

        if (string.IsNullOrWhiteSpace(resolvedPath))
        {
            error = "Path is empty";
            return false;
        }

        // Check if path exists
        if (!Directory.Exists(resolvedPath) && !File.Exists(resolvedPath))
        {
            error = $"Path does not exist: {resolvedPath}";
            return false;
        }

        // Validate path is within vault (security check)
        var normalizedPath = Path.GetFullPath(resolvedPath).Replace('\\', '/');
        var normalizedVaultRoot = Path.GetFullPath(_vaultRootPath).Replace('\\', '/');

        if (!normalizedPath.StartsWith(normalizedVaultRoot, StringComparison.OrdinalIgnoreCase))
        {
            error = $"Path is outside the vault root: {resolvedPath}";
            return false;
        }

        return true;
    }

    private static bool IsIndexFile(string filePath)
    {
        var fileName = Path.GetFileName(filePath);
        return fileName.Equals("_index.md", StringComparison.OrdinalIgnoreCase)
            || fileName.Equals("index.md", StringComparison.OrdinalIgnoreCase);
    }

    private static TagOperationResult CreateResult(Dictionary<string, int> stats, bool dryRun, string operation)
    {
        var filesProcessed = stats.GetValueOrDefault("FilesProcessed", 0);
        var filesModified = stats.GetValueOrDefault("FilesModified", 0);
        var tagsAdded = stats.GetValueOrDefault("TagsAdded", 0);
        var filesWithErrors = stats.GetValueOrDefault("FilesWithErrors", 0);

        var dryRunPrefix = dryRun ? "[DRY RUN] " : "";
        var message = $"{dryRunPrefix}{operation}: processed {filesProcessed} files, modified {filesModified}, added {tagsAdded} tags";

        if (filesWithErrors > 0)
        {
            message += $", {filesWithErrors} errors";
        }

        return new TagOperationResult
        {
            Success = filesWithErrors == 0,
            Message = message,
            FilesProcessed = filesProcessed,
            FilesModified = filesModified,
            TagsAdded = tagsAdded,
            FilesWithErrors = filesWithErrors,
            DryRun = dryRun
        };
    }

    private static TagOperationResult CreateErrorResult(string error, bool dryRun)
    {
        return new TagOperationResult
        {
            Success = false,
            Message = error,
            FilesProcessed = 0,
            FilesModified = 0,
            TagsAdded = 0,
            FilesWithErrors = 1,
            DryRun = dryRun,
            ErrorMessage = error
        };
    }

    private static string? GetSuggestedFix(string diagnosticMessage)
    {
        // Provide common fix suggestions based on diagnostic patterns
        if (diagnosticMessage.Contains("missing colon", StringComparison.OrdinalIgnoreCase))
        {
            return "Add a colon after the key name (e.g., 'title: My Title')";
        }

        if (diagnosticMessage.Contains("indentation", StringComparison.OrdinalIgnoreCase))
        {
            return "Use consistent indentation (2 spaces recommended)";
        }

        if (diagnosticMessage.Contains("quote", StringComparison.OrdinalIgnoreCase))
        {
            return "Wrap values containing special characters in quotes";
        }

        if (diagnosticMessage.Contains("frontmatter", StringComparison.OrdinalIgnoreCase))
        {
            return "Ensure frontmatter is enclosed by '---' on separate lines";
        }

        return null;
    }

    #endregion
}

File: src/obsidian-plugin/services/TagService.ts
// Licensed under the MIT License. See LICENSE file in the project root for full license information.

import { App, TFile, TFolder, CachedMetadata } from 'obsidian';
import { TagOperationResult, YamlDiagnosisResult } from '../models';

/**
 * Service interface for tag management operations in the Obsidian vault.
 * Provides high-level API for tag operations designed for integration with the plugin.
 */
export interface ITagService {
  /**
   * Adds nested tags to markdown files based on frontmatter fields.
   * 
   * @param path - Path to process (file or folder).
   * @param dryRun - If true, simulates changes without modifying files.
   * @returns Result containing statistics and any errors.
   */
  addNestedTags(path: string, dryRun?: boolean): Promise<TagOperationResult>;

  /**
   * Consolidates duplicate and similar tags across files.
   * 
   * @param path - Path to process, or null for entire vault.
   * @param dryRun - If true, simulates changes without modifying files.
   * @returns Result containing statistics and any errors.
   */
  consolidateTags(path?: string, dryRun?: boolean): Promise<TagOperationResult>;

  /**
   * Restructures tags according to the configured hierarchy.
   * 
   * @param path - Path to process, or null for entire vault.
   * @param dryRun - If true, simulates changes without modifying files.
   * @returns Result containing statistics and any errors.
   */
  restructureTags(path?: string, dryRun?: boolean): Promise<TagOperationResult>;

  /**
   * Updates or adds a frontmatter key-value pair in markdown files.
   * 
   * @param path - Path to file or directory to process.
   * @param key - Frontmatter key to update.
   * @param value - New value for the key.
   * @param dryRun - If true, simulates changes without modifying files.
   * @returns Result containing statistics and any errors.
   */
  updateFrontmatter(path: string, key: string, value: string, dryRun?: boolean): Promise<TagOperationResult>;

  /**
   * Diagnoses YAML frontmatter issues in markdown files.
   * 
   * @param path - Path to scan, or null for entire vault.
   * @returns Diagnosis result containing files with issues and suggested fixes.
   */
  diagnoseYaml(path?: string): Promise<YamlDiagnosisResult>;

  /**
   * Gets tags from a specific file.
   * 
   * @param filePath - Path to the file.
   * @returns Array of tag strings.
   */
  getTags(filePath: string): Promise<string[]>;

  /**
   * Adds a tag to a file.
   * 
   * @param filePath - Path to the file.
   * @param tag - Tag to add.
   */
  addTag(filePath: string, tag: string): Promise<void>;

  /**
   * Removes a tag from a file.
   * 
   * @param filePath - Path to the file.
   * @param tag - Tag to remove.
   */
  removeTag(filePath: string, tag: string): Promise<void>;
}

/**
 * Implementation of tag service using Obsidian's API.
 */
export class TagService implements ITagService {
  constructor(private readonly app: App) {}

  async addNestedTags(path: string, dryRun = false): Promise<TagOperationResult> {
    // TODO: Implement nested tag addition based on frontmatter fields
    console.log(`[TagService] addNestedTags: path=${path}, dryRun=${dryRun}`);
    
    return {
      success: false,
      message: 'Not yet implemented',
      filesProcessed: 0,
      filesModified: 0,
      tagsAdded: 0,
      filesWithErrors: 0,
      dryRun,
      errorMessage: 'This feature is not yet implemented'
    };
  }

  async consolidateTags(path?: string, dryRun = false): Promise<TagOperationResult> {
    // TODO: Implement tag consolidation
    console.log(`[TagService] consolidateTags: path=${path}, dryRun=${dryRun}`);
    
    return {
      success: false,
      message: 'Not yet implemented',
      filesProcessed: 0,
      filesModified: 0,
      tagsAdded: 0,
      filesWithErrors: 0,
      dryRun,
      errorMessage: 'This feature is not yet implemented'
    };
  }

  async restructureTags(path?: string, dryRun = false): Promise<TagOperationResult> {
    // TODO: Implement tag restructuring
    console.log(`[TagService] restructureTags: path=${path}, dryRun=${dryRun}`);
    
    return {
      success: false,
      message: 'Not yet implemented',
      filesProcessed: 0,
      filesModified: 0,
      tagsAdded: 0,
      filesWithErrors: 0,
      dryRun,
      errorMessage: 'This feature is not yet implemented'
    };
  }

  async updateFrontmatter(path: string, key: string, value: string, dryRun = false): Promise<TagOperationResult> {
    console.log(`[TagService] updateFrontmatter: path=${path}, key=${key}, value=${value}, dryRun=${dryRun}`);
    
    try {
      const file = this.app.vault.getAbstractFileByPath(path);
      
      if (!file) {
        return {
          success: false,
          message: `File not found: ${path}`,
          filesProcessed: 0,
          filesModified: 0,
          tagsAdded: 0,
          filesWithErrors: 1,
          dryRun,
          errorMessage: `File not found: ${path}`
        };
      }

      if (file instanceof TFile && file.extension === 'md') {
        if (!dryRun) {
          await this.app.fileManager.processFrontMatter(file, (frontmatter) => {
            frontmatter[key] = value;
          });
        }
        
        return {
          success: true,
          message: `Updated frontmatter ${key}=${value} in ${path}`,
          filesProcessed: 1,
          filesModified: dryRun ? 0 : 1,
          tagsAdded: 0,
          filesWithErrors: 0,
          dryRun
        };
      } else if (file instanceof TFolder) {
        // TODO: Process all markdown files in folder
        return {
          success: false,
          message: 'Folder processing not yet implemented',
          filesProcessed: 0,
          filesModified: 0,
          tagsAdded: 0,
          filesWithErrors: 0,
          dryRun,
          errorMessage: 'Folder processing not yet implemented'
        };
      }

      return {
        success: false,
        message: `Not a markdown file: ${path}`,
        filesProcessed: 0,
        filesModified: 0,
        tagsAdded: 0,
        filesWithErrors: 1,
        dryRun,
        errorMessage: `Not a markdown file: ${path}`
      };
    } catch (error) {
      return {
        success: false,
        message: `Error updating frontmatter: ${error}`,
        filesProcessed: 0,
        filesModified: 0,
        tagsAdded: 0,
        filesWithErrors: 1,
        dryRun,
        errorMessage: String(error)
      };
    }
  }

  async diagnoseYaml(path?: string): Promise<YamlDiagnosisResult> {
    // TODO: Implement YAML diagnosis
    console.log(`[TagService] diagnoseYaml: path=${path}`);
    
    return {
      success: false,
      message: 'Not yet implemented',
      filesScanned: 0,
      filesWithIssues: 0,
      issues: []
    };
  }

  async getTags(filePath: string): Promise<string[]> {
    const file = this.app.vault.getAbstractFileByPath(filePath);
    
    if (!(file instanceof TFile)) {
      return [];
    }

    const metadata = this.app.metadataCache.getFileCache(file);
    
    if (!metadata) {
      return [];
    }

    // Get tags from frontmatter
    const frontmatterTags = metadata.frontmatter?.tags || [];
    const tags = Array.isArray(frontmatterTags) ? frontmatterTags : [frontmatterTags];
    
    // Get tags from content (inline tags)
    const inlineTags = metadata.tags?.map(tag => tag.tag) || [];
    
    // Combine and deduplicate
    return [...new Set([...tags, ...inlineTags])];
  }

  async addTag(filePath: string, tag: string): Promise<void> {
    const file = this.app.vault.getAbstractFileByPath(filePath);
    
    if (!(file instanceof TFile) || file.extension !== 'md') {
      throw new Error(`Not a markdown file: ${filePath}`);
    }

    await this.app.fileManager.processFrontMatter(file, (frontmatter) => {
      if (!frontmatter.tags) {
        frontmatter.tags = [];
      } else if (!Array.isArray(frontmatter.tags)) {
        frontmatter.tags = [frontmatter.tags];
      }
      
      if (!frontmatter.tags.includes(tag)) {
        frontmatter.tags.push(tag);
      }
    });
  }

  async removeTag(filePath: string, tag: string): Promise<void> {
    const file = this.app.vault.getAbstractFileByPath(filePath);
    
    if (!(file instanceof TFile) || file.extension !== 'md') {
      throw new Error(`Not a markdown file: ${filePath}`);
    }

    await this.app.fileManager.processFrontMatter(file, (frontmatter) => {
      if (Array.isArray(frontmatter.tags)) {
        frontmatter.tags = frontmatter.tags.filter((t: string) => t !== tag);
      }
    });
  }
}
